enhancements/testing/improved-platform-tests.md

```diff
@@ -0,0 +1,412 @@
+        else external test?
+            rect rgb(177, 177, 177)
+                O->>+E: run #lt;test name#gt;
+                E->>+O: results
+            end
+        end
+    end
+    O->>+U: results
+-->
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+To achieve the first step of the proposal, we need to:
+1. Provide a library for building `<binary>-tests` which exposes two commands:
+   - `list` - responsible for listing tests in JSON format;
+   - `run` - run a single test, returning results in ginkgo compatible format;
+2. Provide an image annotation or similar mechanism informing where the test
+   binary is embedded in the release payload.
+
+The proposed extensions to `openshift-tests` will require the following changes:
+1. Add extraction logic responsible for retrieving test binaries from the
+   release payload, based on the annotations provided by the authors of these
+   images.
+2. Add logic responsible for iterating over all test binaries invoking the `list`
+   subcommand, and reading the list of tests provided by the binary. Every tests
+   should have an associated binary name, which will be used during invocation.
+3. Expand the current filtering, shuffling, and randomization logic to also take
+   into account the new tests coming from external binaries.
+4. Add logic responsible for iterating over all tests, invoking the `run`
+   subcommand using the associated test's binary and reading its results.
+5. Provide a fallback path allowing to use the vendored tests, only when either
+   the release image is not available, or on demand (`OPENSHIFT_SKIP_EXTERNAL_TESTS`
+   environment variable, for example).
+
+### Risks and Mitigations
+
+- The current policy when the [origin](https://github.com/openshift/origin/) pull
+  request bringing the updated k8s version breaks the CI is to revert PR. This
+  policy will be harder to execute with this change because reverting the o/k
+  rebase PR can be problematic and was never tried before.
+
+There should be fewer reasons to do any reverting, because with this new mechanism
+we should get the appropriate signal during extensive payload testing required
+to land every k8s bump PR.
+
+- Some of the work that has been done to build the partner certification tool will
+  need to be updated to pull additional images with separate test binaries.
+
+The `openshift-tests` binary will contain the necessary logic how to retrieve those
+additional tests binaries from the release payload, so even in the offline installation
+as long as the release payload is reachable, the tests should continue working.
+
+- Not available release image can prevent from running the tests.
+
+The proposed mechanism will contain a fallback path, which will ensure that in
+cases where the release payload is not reachable we will be able to run all the
+tests bundled within the `openshift-tests` binary. This includes the basic
+OpenShift and Kubernetes conformance tests.
```

Comment:
long term i don't see us keeping the k8s tests in origin, that would mean extra rebase work.

also if there is a fallback path, it needs to be opt-in and very clear that it's been taken, because it means a significant change in test coverage when falling back.  (To the point that i'm not even sure we'd ever want automatic fallback, even with a "--allow-fallback" argument.  I'd rather setup two jobs, one which uses external content and one which doesn't, rather than have a single job that sometimes falls back and sometimes doesn't)

Reply:
I'm willing to remove the automatic fallback, since like you're describing this will force you to be explicit about running vendored k8s vs external.
As for your worry about diverging tests, I'm less worried about that, major bulk of test changes is coming from the major k8s bump, even though patch releases have some fixes, they are rather cosmetic from what I've seen over the years.

Reply:
https://github.com/openshift/origin/pull/28000 removes the automatic fallback. I'll update the proposal accordingly.